libdrgn/program.c

```diff
@@ -860,6 +901,17 @@ drgn_program_cache_core_dump_notes(struct drgn_program *prog)
 		assert(it.entry);
 		prog->crashed_thread = it.entry;
 	}
+
+	if (found_main_thread) {
+		/*
+		 * Similar to the above, but now caching the main thread
+		 */
+		struct drgn_thread_set_iterator it =
+			drgn_thread_set_search(&prog->thread_set,
+					       &main_thread_tid);
+		assert(it.entry);
```

Comment:
This assertion will fail if the core dump is malformed and doesn't have a `PRSTATUS` note matching the `PRPSINFO` note. We shouldn't crash on bad input. Remove this `assert` and the `if (!prog->main_thread)` check in `drgn_program_main_thread()` will take care of it.

(The assertion for `crashed_thread` is different: we know we inserted a thread into the set with that TID, so we must be able to find it.)

Reply:
Thanks for the explanation, makes sense! Fixed